src/py_adaptors.h

```diff
@@ -177,47 +264,90 @@ class PathIterator
     {
-        return (void *)m_vertices;
+        return (void *)m_vertices.data();
     }
 };
 
 class PathGenerator
 {
-    PyObject *m_paths;
     Py_ssize_t m_npaths;
+    bool m_is_optimized;
 
-  public:
-    typedef PathIterator path_iterator;
+    // Used for optimized path collections
+    StridedMemory3D<double> m_vertices;
+    StridedMemory2D<uint8_t> m_codes;
+    unsigned m_path_length;
+    bool m_should_simplify;
+    double m_simplify_threshold;
 
-    PathGenerator(PyObject *obj) : m_paths(NULL), m_npaths(0)
-    {
-        if (!set(obj)) {
-            throw py::exception();
-        }
-    }
+    // Used for general sequences
+    PyObject *m_paths;
 
-    ~PathGenerator()
-    {
-        Py_XDECREF(m_paths);
-    }
+  public:
+    typedef PathIterator path_iterator;
 
-    int set(PyObject *obj)
+    PathGenerator(PyObject *obj)
+        : m_npaths(0),
+          m_is_optimized(false),
+          m_vertices(),
+          m_codes(),
+          m_path_length(0),
+          m_should_simplify(false),
+          m_simplify_threshold(0),
+          m_paths(NULL)
     {
         if (!PySequence_Check(obj)) {
-            return 0;
+            throw py::exception();
         }
 
         m_paths = obj;
         Py_INCREF(m_paths);
 
         m_npaths = PySequence_Size(m_paths);
 
-        return 1;
+        PyObject *is_uniform_obj = PyObject_GetAttrString(obj, "_is_uniform_path_collection");
+        PyErr_Clear(); // The attribute might not be there.
+        m_is_optimized = is_uniform_obj == Py_True;
+        Py_XDECREF(is_uniform_obj);
+        if (m_is_optimized) {
+            PyArrayObject *vertices_obj = (PyArrayObject*)PyObject_GetAttrString(obj, "_vertices");
+            PyArrayObject *codes_obj = (PyArrayObject*)PyObject_GetAttrString(obj, "_codes");
+            PyObject *should_simplify_obj = PyObject_GetAttrString(obj, "should_simplify");
+            PyObject *simplify_threshold_obj = PyObject_GetAttrString(obj, "simplify_threshold");
+            if (vertices_obj == NULL || codes_obj == NULL ||
+                should_simplify_obj == NULL || simplify_threshold_obj == NULL) {
+                PyErr_SetString(PyExc_ValueError, "Expected a uniform path collection");
+                goto end;
+            }
+            if (!PyArray_Check(vertices_obj) || !PyArray_Check(codes_obj)) {
+                PyErr_SetString(PyExc_ValueError, "Vertices and codes should be NumPy arrays");
+                goto end;
+            }
+            m_vertices = StridedMemory3D<double>(vertices_obj);
+            m_codes = StridedMemory2D<uint8_t>(codes_obj);
+            m_path_length = PyArray_DIM(vertices_obj, 1);
+            m_should_simplify = should_simplify_obj == Py_True;
+            m_simplify_threshold = PyFloat_AsDouble(simplify_threshold_obj);
+end:
+            Py_XDECREF(vertices_obj);
+            Py_XDECREF(codes_obj);
+            Py_XDECREF(should_simplify_obj);
+            Py_XDECREF(simplify_threshold_obj);
```

Comment:
The more I look at this code the more I want to write a RAII wrapper for managing Python refcounts. I've seen some potential memory leaks in the `PathIterator` code before changes too, so it might be useful to start using it in the whole codebase instead of dealing with it all manually. `pybind11` is quite standard for that too, but it requires C++11...

Reply:
I agree pybind11 would be nice, I'll mention this during the dev call, this may save a lot of energy on reviewing this...

Reply:
Well you don't need all of pybind here, so I could easily write RAII wrappers for refcount management. I just don't want to make this patch too large and it seems like it should undergo a larger discussion. This code is definitely more consistent with the rest of the code for now.

Reply:
I think even a trivial RAII wrapper (basically the equivalent of `pybind11::object`) would be nice as a separate PR?

Reply:
I think so. Would you rather merge the RAII thing before this?

Reply:
yes